store/tikv/region_cache.go

```diff
@@ -1242,12 +1243,30 @@ func (c *RegionCache) getProxyStore(region *Region, store *Store, rs *RegionStor
 	}
 
 	tikvNum := rs.accessStoreNum(TiKVOnly)
-	for index := 0; index < tikvNum; index++ {
+	if tikvNum <= 1 {
+		return
+	}
+
+	// Randomly select an non-leader peer
+	first := rand.Intn(tikvNum - 1)
+	if first >= int(workStoreIdx) {
+		first = (first + 1) % tikvNum
+	}
+
+	// If the current selected peer is not reachable, switch to the next one, until a reachable peer is found or all
+	// peers are checked.
+	for i := 0; i < tikvNum; i++ {
+		index := (i + first) % tikvNum
 		// Skip work store which is the actual store to be accessed
 		if index == int(workStoreIdx) {
 			continue
 		}
 		storeIdx, store := rs.accessStore(TiKVOnly, AccessIndex(index))
+		// Skip unreachable stores.
+		if atomic.LoadInt32(&store.needForwarding) != 0 {
+			continue
+		}
+
```

Comment:
I think this function shouldn't report errors. If it reports, the error is returned to the user directly.